service/header_service.go

```diff
@@ -291,6 +292,66 @@ func (hs *HeaderService) GetMerkleRootsConfirmations(
 	return hs.repo.Headers.GetMerkleRootsConfirmations(request, hs.merkleCfg.MaxBlockHeightExcess)
 }
 
+// LocateHeadersGetHeaders returns headers with given hashes.
+func (hs *HeaderService) LocateHeadersGetHeaders(locators []*chainhash.Hash, hashstop *chainhash.Hash) []*wire.BlockHeader {
+	headers := hs.locateHeadersGetHeaders(locators, hashstop)
+	return headers
+}
+
+func (hs *HeaderService) locateHeadersGetHeaders(locators []*chainhash.Hash, hashstop *chainhash.Hash) []*wire.BlockHeader {
+	hashes := make([]string, len(locators))
+	for i, v := range locators {
+		hashes[i] = v.String()
+	}
+
+	startHeight, err := hs.repo.Headers.GetHeadersStartHeight(hashes)
+	if err != nil {
+		log.Trace().Msgf("Error getting headers of locators: %v", err)
+		return nil
+	}
+	var stopHeight int
+	if hashstop.IsEqual(&chainhash.Hash{}) {
+		stopHeight = startHeight + wire.MaxCFHeadersPerMsg
+	} else {
+		stopHeight, err = hs.repo.Headers.GetHeadersStopHeight(hashstop.String())
+		if err != nil {
+			log.Trace().Msgf("Error getting hashstop height: %v", err)
+			return nil
+		}
+	}
+	// Check if hashStop is lower than first valid height
+	if stopHeight <= startHeight {
+		log.Trace().Msgf("HashStop is lower than first valid height")
+		return nil
+	}
+
+	// Check if peer requested number of headers is higher than the maximum number of headers per message
+	if wire.MaxCFHeadersPerMsg < stopHeight-startHeight {
+		stopHeight = startHeight + wire.MaxCFHeadersPerMsg
+	}
+
+	dbHeaders, err := hs.repo.Headers.GetHeadersByHeightRange(startHeight+1, stopHeight)
+	if err != nil {
+		log.Trace().Msgf("Error getting headers between heights: %v", err)
```

Comment:
maybe it's good idea to log it as Error, because this error will rather suggest that we have some problem with db.

Reply:
changed to handle errors log at the peer side